### PR TITLE
Threaded comment moderation: fix restoring unmoderated comments

### PR DESCRIPTION
**Pull request by ScoutHarris**
Ref: #17629, #17758, p5T066-2Wh#comment-11057

This fixes a couple of issues with moderating post comments from the threaded comments view.

1) If the first comment was moderated (removing it from the list), the post details Comments snippet was not updated (i.e. the old first comment was still displayed).

I've fixed this by posting a `NSNotification` so the post details is aware of any comment changes.

2) When a comment was moderated (removing it from the list), the Comment table would show an empty comment row. 

screen-shot-2022-01-26-at-12 31 copy

The cause was that, to facilitate `Undo`ing comment moderation, the comment was kept in the cache. However, because it was in the cache, the table's `fetchedObjects` still contained it, thus the table tried to show it.

I've resolved this by restoring the `Comment` when the user selects `Undo`.
- When a comment is moderated, a copy of the Comment is stored.
- If the user selects `Undo` on the confirmation notice, the `Comment` is recreated, and is then re-Approved.

To test:

**Updating Comments snippet:**
- Go to the Reader. Select a post with comments and that you have permission to moderate comments on.
- Note the comment displayed in the Comments snippet.
- Go to the Comments view.
- On the first comment, tap the ellipsis button and select any status. 

Screen Shot 2022-01-27 at 4 45 20 PM

- Go back to the post details.
- Verify the Comments snippet shows the next top level comment.

**Undoing comment moderation:**
- Go to the Reader. Select a post with comments and that you have permission to moderate comments on.
- Go to the Comments view.
- On any comment, tap the ellipsis button and select any status. 
- Verify the comment is removed from the view, and no empty comments are displayed (as shown above).
- Tap `Undo` on the confirmation notice.
- Verify the comment is restored and has the same hierarchy as before.

## Regression Notes
1. Potential unintended areas of impact

2. What I did to test those areas of impact (or what existing automated tests I relied on)

3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by ScoutHarris** [comment]
@frosty can I bother you for a review please? Thank you! 🙇

File: WordPress/Classes/Services/CommentService.m [hunk 1]
```diff
@@ -1119,6 +1119,44 @@ - (void)updateCommentAndSave:(Comment *)comment withRemoteComment:(RemoteComment
...
+    }
+    
+    // Otherwise create a new Comment.
+    NSString *entityName = NSStringFromClass([Comment class]);
+    comment = [NSEntityDescription insertNewObjectForEntityForName:entityName inManagedObjectContext:self.managedObjectContext];
+    
+    [self updateComment:comment withRemoteComment:remoteComment];
+    comment.post = post;
+    comment.depth = depth;
+    comment.hierarchy = hierarchy;
```
Comment (ScoutHarris) [inline WordPress/Classes/Services/CommentService.m:1156]: The hierarchy isn't calculated correctly with the Comment in isolation (i.e. without it's ancestors). It is calculated correctly when the full hierarchical comments are refreshed, but to get the comment to display properly when restored, the depth and hierarchy need to be provided.

**Comment by ScoutHarris** [comment]
@mokagio - FYI - this will need to go in the next WPiOS 19.1 release when merged. Thank you!

File: WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.h [hunk 2]
```diff
@@ -31,4 +31,12 @@ typedef NS_ENUM(NSUInteger, ReaderCommentsSource) {
 /// Navigates to the specified comment when the view appears
 @property (nonatomic, strong) NSNumber *navigateToCommentID;
 
+
+
+// Undo comment moderation support.
+- (void)refreshTableViewAndNoResultsView;
```
Comment (frosty) [inline WordPress/Classes/ViewRelated/Reader/Comments/ReaderCommentsViewController.h]: Could we move the method declaration below the properties?

**Review by frosty — COMMENTED** [review]
Code looks fine to me 👍 

The first test, **updating comments snippet** worked fine for me, but the second (**undoing comment moderation**) did not. When I undid the operation, a comment would end up missing from the list until I left the screen and then returned to it. Here's the sequence of events:

| 3 comments | Middle comment set to unapproved | Undo tapped, bottom comment has gone |
|---|---|---|
| Simulator Screen Shot - iPhone 13 Pro - 2022-01-28 at 13 17 12 | Simulator Screen Shot - iPhone 13 Pro - 2022-01-28 at 13 17 18 | Simulator Screen Shot - iPhone 13 Pro - 2022-01-28 at 13 17 23 |

**Comment by ScoutHarris** [comment]
Hey @frosty . I think I got it this time. I had failed to specify the siteID when fetching the post. I've [corrected](https://github.com/wordpress-mobile/WordPress-iOS/pull/17834/commits/c5b33c63efdc318980e9deeadcac2a76f94896e3) that. So restored comments should now re-appear correctly. Ready for another review please. Thank you!

**Review by frosty — COMMENTED** [review]
I'm afraid undoing still doing seem to be quite right for me. When I undo I'm not seeing the comment return, and in fact I'm also seeing the below comment disappear:

https://user-images.githubusercontent.com/4780/151790269-114fb8f8-0d2c-44a3-b478-0d8be911f5ae.mov

**Comment by ScoutHarris** [comment]
Hey @frosty !

Thanks for noticing that! It took several attempts, but I did finally see it. Not sure how it didn't happen all the time, but I'm glad it at least happened to you! 😬 

It looks like it was caused by the post's `commentCount` property being out of sync with the actual number of comments (`post.comments` array). So I simplified updating the `commentCount` property. That seems to have resolved the missing comment. 🤞 this does it!

Ready for another go please kind sir!

**Review by frosty — COMMENTED** [review]
Hmm, it seems as though the Undo action just doesn't seem to be working for me (sorry!). Here's a video where I show a post that has 2 approved comments. I unapprove one, and then try to undo the action. 

The comment doesn't return, and the Reader post view now says 1 comment. If I check in the Comments screen for the site, it now shows just one approved post.

https://user-images.githubusercontent.com/4780/151968006-c8f6bc98-baec-4b75-af4c-914b57ca5a7b.mov

**Comment by ScoutHarris** [comment]
Hey @frosty .

Clearly I'm missing something. I'm going to table this for now and come back to it for the next release. I've created  #17858 to disable the feature in 19.1.

Sorry for the bother and thank you for all the testing!

@mokagio - this will not go in 19.1, but #17858 will. Thank you!

**Pull request closed without merging**